test/vm_test.go

```diff
@@ -0,0 +1,260 @@
+	defer vm.Close(t)
+	require.NotNil(t, vm)
+
+	vm.AddSSH(t, "tcp")
+
+	tempDir := t.TempDir()
+	vsockListenPath := filepath.Join(tempDir, "vsock-listen.sock")
+	ln, err := net.Listen("unix", vsockListenPath)
+	require.NoError(t, err)
+	go func() {
+		conn, err := ln.Accept()
+		// call ln.Close() after a timeout to unblock Accept() and fail the test?
+		require.NoError(t, err)
+		data, err := io.ReadAll(conn)
+		require.NoError(t, err)
+		log.Infof("read %v", string(data))
+		require.Equal(t, []byte("hello host"), data)
+	}()
+	log.Infof("path to vsock socket: %s", vsockListenPath)
+	dev, err := config.VirtioVsockNew(1235, vsockListenPath, true)
+	require.NoError(t, err)
+	vm.AddDevice(t, dev)
+
+	vm.Start(t)
+	vm.WaitForSSH(t)
+
+	vm.SSHRun(t, "echo -n 'hello host' | socat -T 2 STDIN VSOCK-CONNECT:2:1235")
+
+	vm.Stop(t)
+}
+
+func TestFileSharing(t *testing.T) {
+	puipuiProvider := NewPuipuiProvider()
+	log.Info("fetching os image")
+	tempDir := t.TempDir()
+	err := puipuiProvider.Fetch(tempDir)
+	require.NoError(t, err)
+
+	vm := NewTestVM(t, puipuiProvider)
+	defer vm.Close(t)
+	require.NotNil(t, vm)
+
+	vm.AddSSH(t, "tcp")
+
+	sharedDir := t.TempDir()
+	share, err := config.VirtioFsNew(sharedDir, "vfkit-test-share")
+	require.NoError(t, err)
+	vm.AddDevice(t, share)
+	log.Infof("shared directory: %s", sharedDir)
+
+	vm.Start(t)
+	vm.WaitForSSH(t)
+
+	vm.SSHRun(t, "mkdir /mnt")
+	vm.SSHRun(t, "mount -t virtiofs vfkit-test-share /mnt")
+
+	err = os.WriteFile(filepath.Join(sharedDir, "from-host.txt"), []byte("data from host"), 0600)
+	require.NoError(t, err)
+	data, err := vm.SSHCombinedOutput(t, "cat /mnt/from-host.txt")
+	require.NoError(t, err)
+	require.Equal(t, "data from host", string(data))
+
+	vm.SSHRun(t, "echo -n 'data from guest' > /mnt/from-guest.txt")
+	data, err = os.ReadFile(filepath.Join(sharedDir, "from-guest.txt"))
+	require.NoError(t, err)
+	require.Equal(t, "data from guest", string(data))
+
+	vm.Stop(t)
+}
+
+type createDevFunc func(t *testing.T) (config.VirtioDevice, error)
+type pciidTest struct {
+	vendorID  int
+	deviceID  int
+	createDev createDevFunc
+}
+
+var pciidTests = map[string]pciidTest{
+	"virtio-net": {
+		vendorID: 0x1af4, // Red Hat
+		deviceID: 0x1041,
+		createDev: func(t *testing.T) (config.VirtioDevice, error) {
+			return config.VirtioNetNew("")
+		},
+	},
+	"virtio-serial": {
+		vendorID: 0x1af4, // Red Hat
+		deviceID: 0x1043,
+		createDev: func(t *testing.T) (config.VirtioDevice, error) {
+			return config.VirtioSerialNew(filepath.Join(t.TempDir(), "serial.log"))
+		},
+	},
+	"virtio-rng": {
+		vendorID: 0x1af4, // Red Hat
+		deviceID: 0x1044,
+		createDev: func(t *testing.T) (config.VirtioDevice, error) {
+			return config.VirtioRngNew()
+		},
+	},
+	"virtio-fs": {
+		vendorID: 0x1af4, // Red Hat
+		deviceID: 0x105a,
+		createDev: func(t *testing.T) (config.VirtioDevice, error) {
+			return config.VirtioFsNew("./", "vfkit-share-test")
+		},
+	},
+	"virtio-gpu": {
```

Comment:
`virtio-gpu`, `virtio-input/*` should only be run on macOS 13+ as these devices are not supported with earlier virtualization framework versions.